src/unix/mediactrl_gstplayer.cpp

```diff
@@ -451,12 +494,22 @@ wxLongLong wxGStreamerMediaBackend::GetDuration()
     return GST_CLOCK_TIME_IS_VALID(duration) ? duration / GST_MSECOND : 0;
 }
 
-void wxGStreamerMediaBackend::Move(int WXUNUSED(x), int WXUNUSED(y), int WXUNUSED(w), int WXUNUSED(h))
+void wxGStreamerMediaBackend::Move(int WXUNUSED(x), int WXUNUSED(y), int w, int h)
 {
-    /* Nothing to be done here, at least for GTK+. For other toolkits we might
-     * have to call
-     * gst_player_video_overlay_video_renderer_set_render_rectangle() here
+    /* GTK+ on X11 does not require explicitely setting the video size.
+     * Other toolkits might need to do something similar to GTK wayland too.
      */
+#ifdef __WXGTK3__
+    GdkWindow* window = gtk_widget_get_window(m_ctrl->m_wxwindow);
+    if (window && strcmp("GdkWaylandWindow", g_type_name(G_TYPE_FROM_INSTANCE(window))) == 0) {
+        int x, y;
+        GtkAllocation allocation;
+        gdk_window_get_origin(window, &x, &y);
+	gtk_widget_get_allocation(m_ctrl->m_wxwindow, &allocation);
+	gst_player_video_overlay_video_renderer_set_render_rectangle(GST_PLAYER_VIDEO_OVERLAY_VIDEO_RENDERER(m_video_renderer),
+                                                                         x, y, w, h);
```

Comment:
You don't ever use the result of `gtk_widget_get_allocation()`.  Also, the indentation is inconsistent here.

Reply:
Gah, thanks. That's what I get for copy-pasta and originally left it to check it matched the function arguments then only removed the debug print :/
Force pushed removing allocation here and fixing tabs in both commits.

Reply:
Thanks, that was fast. :-)  Working on testing this on my system too.

Reply:
cheers, more tests are more than welcome!
Actually thinking back on the allocation the function argument happened to be the same, but Move width/height would be the full window dimensions and we want the widget's width/height; and they just happened to be the same for the mediaplayer example? In which case we'd want to get allocation like the other codepath.

Looking at mediaplayer.cpp I don't get why the video goes out of the m_mediactrl box that should be left of playlist/above controls and draws all the NotebookPage, given that's what it does on gtk3/X11 I didn't think too hard about it but that might not be correct?

Reply:
It worked for me also, but I had to set `GST_PLUGIN_FEATURE_RANK=waylandsink:257` like you did.  Otherwise, it crashes.  Is there any way we can configure it to use the waylandsink automatically on Wayland?

Reply:
Hm, no - I was setting that variable to test on my laptop, but it should work with other backends.. Or well at least glimagesink that my laptop uses by default. I wasn't aware of any crash left.
Ultimately it was also working with glimagesink on the imx board too, it's just that the board somehow prefers waylandsink by default (and that sink dropped frames on my test so I'm actually not so happy about it...) while my laptop prefers glimagesink and I have no idea where this preference happens... Looking at gst-inspect-1.0 something is overriding it but I'm not sure what (first is laptop, second is board):
```
$ gst-inspect-1.0 -a | grep -E '(waylandsink|glimagesink).*Rank'
waylandsink:   Rank                     marginal (64)
glimagesink:   Rank                     secondary (128)
$ gst-inspect-1.0 -a | grep -E '(waylandsink|glimagesink).*Rank'
waylandsink:   Rank                     primary + 2 (258)
glimagesink:   Rank                     secondary (128)
```
Ah, it's in the code: https://source.codeaurora.org/external/imx/gst-plugins-bad/commit/?h=MM_04.03.02_1710_L4.9.51_MX8M_BETA&id=0c9a5ffedf6c8d26b080dad181737cabb6c5bff9

Well, doesn't help us with config files :|


Anyway, it shouldn't crash -- could you check what sink your environment uses by default?
I'm not 100% certain of a good way of checking, `GST_DEBUG=4` should output hints, but lately I found out about `GST_DEBUG_DUMP_DOT_DIR` which if set will generate a graphviz file that shows the plugins involved if it doesn't crash before that.

Reply:
So, it's trying to use `xvimagesink` so I guess it's no surprise that it crashes.

Reply:
hm it should detect that xvimagesink is not available when probing it; I'm getting this kind of logs with GST_DEBUG=4
```
0:00:00.042219541 1195814 0x5644c2a45f00 INFO     GST_ELEMENT_FACTORY gstelementfactory.c:363:gst_element_factory_create: creating element "xvimagesink"
0:00:00.042358579 1195814 0x5644c2a45f00 INFO        GST_ELEMENT_PADS gstelement.c:752:gst_element_add_pad:<GstBaseSink@0x7f384c0692a0> adding pad 'sink'
0:00:00.042442401 1195814 0x5644c2a45f00 INFO        GST_ERROR_SYSTEM gstelement.c:2234:gst_element_message_full_with_details:<xvimagesink0> posting message: Could not initialise Xv output
0:00:00.042457821 1195814 0x5644c2a45f00 INFO        GST_ERROR_SYSTEM gstelement.c:2261:gst_element_message_full_with_details:<xvimagesink0> posted error message: Could not initialise Xv output
0:00:00.042465541 1195814 0x5644c2a45f00 INFO              GST_STATES gstelement.c:3054:gst_element_change_state:<xvimagesink0> have FAILURE change_state return
0:00:00.042472464 1195814 0x5644c2a45f00 INFO              GST_STATES gstelement.c:2641:gst_element_abort_state:<xvimagesink0> aborting state from NULL to READY
0:00:00.042490663 1195814 0x5644c2a45f00 INFO         GST_REFCOUNTING gstelement.c:3345:gst_element_dispose:<xvimagesink0> 0x7f384c0692a0 dispose
0:00:00.042498134 1195814 0x5644c2a45f00 INFO        GST_ELEMENT_PADS gstelement.c:868:gst_element_remove_pad:<xvimagesink0> removing pad 'sink'
0:00:00.042511730 1195814 0x5644c2a45f00 INFO         GST_REFCOUNTING gstelement.c:3390:gst_element_dispose:<xvimagesink0> 0x7f384c0692a0 parent class dispose
0:00:00.042520408 1195814 0x5644c2a45f00 INFO         GST_REFCOUNTING gstelement.c:3422:gst_element_finalize:<xvimagesink0> 0x7f384c0692a0 finalize
0:00:00.042526434 1195814 0x5644c2a45f00 INFO         GST_REFCOUNTING gstelement.c:3427:gst_element_finalize:<xvimagesink0> 0x7f384c0692a0 finalize parent
0:00:00.042532788 1195814 0x5644c2a45f00 WARN                 playbin gstplaybin2.c:4755:autoplug_select_cb:<playbin> Could not activate sink xvimagesink
```

So it's harmless on my setup, there must be something different somewhere.. Ah, I get it to abort if DISPLAY is also set, so that broke setups with Xwayland enabled!!

```
0:00:00.124537653 1196511 0x7f7bd8006f00 INFO              GST_STATES gstelement.c:2769:gst_element_continue_state:<xvimagesink0> completed state change to READY
0:00:00.124583501 1196511 0x7f7bd8006f00 INFO     GST_ELEMENT_FACTORY gstelementfactory.c:360:gst_element_factory_create: creating element "bin" named "vbin"
X Error of failed request:  BadWindow (invalid Window parameter)
  Major opcode of failed request:  3 (X_GetWindowAttributes)
  Resource id in failed request:  0xfb79bb40
  Serial number of failed request:  34
  Current serial number in output stream:  35
```

Not sure how to go about fixing that now, if the window is wayland-native then unset DISPLAY from environment?

Reply:
Alright, I've added a commit that unsets DISPLAY if the main window is wayland.. I'm not sure if we could do better (tell gstreamer not to try certain plugins?), but it works.

Since that required wayland-specific code in CreateControl I also made that we only connect the bus_sync_handler in the wayland case, that way the change definitely won't have any impact on native X11 windows (added this to first commit)